sys/signal.c

```diff
@@ -0,0 +1,221 @@
+    [SIGFPE] = "SIGFPE",   [SIGSEGV] = "SIGSEGV", [SIGKILL] = "SIGKILL",
+    [SIGTERM] = "SIGTERM", [SIGCHLD] = "SIGCHLD", [SIGUSR1] = "SIGUSR1",
+    [SIGUSR2] = "SIGUSR2",
+};
+
+void sighand_init() {
+  /* TODO: A lot of kernel components have a init function whose only purpose is
+     to perform a kmalloc_init. Maybe we could collect them with a linker set,
+     and initialize all pools together? */
+  kmalloc_init(sig_pool, 2, 2);
+}
+
+sighand_t *sighand_new() {
+  sighand_t *sh = kmalloc(sig_pool, sizeof(sighand_t), M_ZERO);
+  mtx_init(&sh->sh_mtx, MTX_RECURSE);
+  sh->sh_refcount = 1;
+  return sh;
+}
+static void sighand_free(sighand_t *sh) {
+  assert(sh->sh_refcount == 0);
+  /* No need to lock the mutex, we have the last existing reference. */
+  kfree(sig_pool, sh);
+}
+
+sighand_t *sighand_copy(sighand_t *sh) {
+  sighand_t *new = sighand_new();
+  mtx_scoped_lock(&sh->sh_mtx);
+  memcpy(new->sh_actions, sh->sh_actions, sizeof(sigaction_t) * SIG_LAST);
+  return new;
+}
+
+void sighand_ref(sighand_t *sh) {
+  mtx_scoped_lock(&sh->sh_mtx);
+  sh->sh_refcount++;
+}
+void sighand_unref(sighand_t *sh) {
+  mtx_scoped_lock(&sh->sh_mtx);
+  if (--sh->sh_refcount == 0)
+    sighand_free(sh);
+}
+
+int do_kill(tid_t tid, int sig) {
+  thread_t *target = thread_get_by_tid(tid);
+  if (target == NULL)
+    return -EINVAL;
+
+  return signal(target, sig);
+}
+
+int do_sigaction(int sig, const sigaction_t *act, sigaction_t *oldact) {
+  thread_t *td = thread_self();
+  /* No need to acquire thread lock, we only look up a reference to the signal
+   * handler structure. */
+  sighand_t *sh = td->td_sighand;
+
+  if (sig < 0 || sig >= SIG_LAST)
+    return -EINVAL;
+
+  if (sig == SIGKILL)
+    return -EINVAL;
+
+  mtx_scoped_lock(&sh->sh_mtx);
+  if (oldact != NULL)
+    memcpy(oldact, &sh->sh_actions[sig], sizeof(sigaction_t));
+
+  memcpy(&sh->sh_actions[sig], act, sizeof(sigaction_t));
+
+  return 0;
+}
+
+static sighandler_t *get_sigact(int sig, sighand_t *sh) {
+  sighandler_t *h = sh->sh_actions[sig].sa_handler;
+  if (h == SIG_DFL)
+    h = signal_default_actions[sig];
+  return h;
+}
+
+int signal(thread_t *target, signo_t sig) {
+
+  /* NOTE: This is a very simple implementation! Unimplemented features:
+     - Thread tracing and debugging
+     - Multiple threads sharing PID (aka: process)
+     - Signal masks
+     - SIGSTOP/SIGCONT
+     These limitations (plus the fact that we currently have very little thread
+     states) make the logic of posting a signal very simple!
+  */
+  assert(sig < SIG_LAST);
+
+  mtx_scoped_lock(&target->td_lock);
+
+  if (target->td_state == TDS_INACTIVE) {
+    /* The thread is already dead. */
+    return -EINVAL;
+  }
+
+  /* If the signal is ignored, don't even bother posting it. */
+  sighand_t *sh = target->td_sighand;
+  if (get_sigact(sig, sh) == SIG_IGN)
+    return 0;
+
+  bit_set(target->td_sigpend, sig);
+
+  /* TODO: If a thread is sleeping, wake it up, so that it continues execution
```

Comment:
The problem is that signal posting IMO should never wake up a sleeping thread, unless one explicitly allowed to do so, by entering `interruptible sleep` state. As I had been browsing `sleepqueue` code I recognized some of its functions are signal-aware, namely: [sleepq_wait_sig](http://bxr.su/FreeBSD/sys/kern/subr_sleepqueue.c#628) and [sleepq_timedwait_sig](http://bxr.su/FreeBSD/sys/kern/subr_sleepqueue.c#666). Indeed, by looking for [TDF_SINTR](http://bxr.su/s?refs=TDF_SINTR&project=FreeBSD) flag, seems like `sleepqueue` is the only component that cares about interruptible property.

When it comes to putting thread to sleep, apparently the most popular interface is [cv_wait_sig](http://bxr.su/s?refs=cv_wait_sig&project=FreeBSD) function and its [variants](http://bxr.su/FreeBSD/sys/sys/condvar.h#77).